### PR TITLE
Simplified TaylorMode implementation

### PR DESCRIPTION
**Pull request by pnkraemer**
# In a Nutshell
Rewrite the core implementation of `odefilter.init_routines.TaylorMode` into a more compact version and explain the background thoroughly.

No more early exits. No more confusing variable names. Transparent tuple/pytree usage. Let me know if something is not clear please!

### CODE REVIEW

File: tests/test_diffeq/test_odefilter/test_init_routines/test_init_routines.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 
 
 @only_if_jax_available
-@pytest.mark.parametrize("num_derivatives", [2, 3, 5])
+@pytest.mark.parametrize("num_derivatives", [0, 1, 2, 3, 5])
```
Comment (pnkraemer) [inline tests/test_diffeq/test_odefilter/test_init_routines/test_init_routines.py:22]: the edge cases (n=0,1) should also be tested.

File: src/probnum/diffeq/odefilter/init_routines/_autodiff.py [hunk 2]
```diff
@@ -202,30 +198,57 @@ class TaylorMode(_AutoDiffBase):
...
+
+    def _taylor_approximation(self, *, f, y0, order):
+        """Compute an `n`th order Taylor approximation of f at y0."""
+        taylor_coefficient_gen = self._taylor_coefficient_generator(f=f, y0=y0)
+
+        # get the nth entry of the coefficient-generator via itertools.islice
+        derivatives = next(itertools.islice(taylor_coefficient_gen, order, None))
+
+        derivatives_as_array = jnp.stack(derivatives, axis=0)
+        return derivatives_as_array
```
[the remark this answers is not in the document]
Comment (nathanaelbosch) [reply src/probnum/diffeq/odefilter/init_routines/_autodiff.py:221]: In a related way, I also don't understand what `islice` does but maybe I could infer that more easily if I understand the output better.

**Review by nathanaelbosch — APPROVED** [review]
Overall the code looks clean. I don't really understand what's happening though so I commented with a few questions. If you want we can reiterate a bit to make this more understandable for badly-informed people like me. If not, feel free to merge :)

File: src/probnum/diffeq/odefilter/init_routines/_autodiff.py [hunk 3]
```diff
@@ -202,30 +198,57 @@ class TaylorMode(_AutoDiffBase):
...
-            (_, [*y0_coeffs_remaining]) = jet(
-                fun=f,
-                primals=(y0,),
-                series=(coeffs_taylor,),
-            )
-            derivs.append(y0_coeffs_remaining[-2][:-1])
-
-        return jnp.asarray(derivs)
+
+        # The `f` parameter is an autonomous ODE vector field that
```
[the remark this answers is not in the document]
Comment (pnkraemer) [reply src/probnum/diffeq/odefilter/init_routines/_autodiff.py:207]: Before entering here, `f` goes through https://github.com/probabilistic-numerics/probnum/blob/4578fe0481de5a8aab070dff6de53e41e4f58767/src/probnum/diffeq/odefilter/init_routines/_autodiff.py#L45
to simplify the autodiff implementation. Does that help?

File: src/probnum/diffeq/odefilter/init_routines/_autodiff.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pnkraemer) [reply src/probnum/diffeq/odefilter/init_routines/_autodiff.py:227]: I elaborated some more in a comment in the code here. Does that help?

File: src/probnum/diffeq/odefilter/init_routines/_autodiff.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nathanaelbosch) [reply src/probnum/diffeq/odefilter/init_routines/_autodiff.py:207]: Thanks, that makes sense

File: src/probnum/diffeq/odefilter/init_routines/_autodiff.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nathanaelbosch) [reply src/probnum/diffeq/odefilter/init_routines/_autodiff.py:227]: Yes, the new comments are helpful to understand what's going on, thanks :)

**Review by nathanaelbosch — APPROVED** [review]
